proto/cacheclient.proto

```diff
@@ -402,16 +402,99 @@ message _ListRemoveRequest {
 
 message _ListRemoveResponse {}
 
+// List fetch + retain index behavior
+// This follows redis behavior for the most part, differences
+// 1. end_index is exclusive, unless Unbounded
+// 2. `start_index` == 0, then we start at the front of list
+// 2. `end_index` == Unbounded, start at the end of the index, inclusive
+/*
+  list_ = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+  # Using the slice notation:
+  # start:end
+
+
+  # start=0, end=Unbounded
+  list_[0:] == [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+
+  # start=5, end=Unbounded
+  list_[5:] == [5, 6, 7, 8, 9]
+
+  # start=5, end=0
+  list_[5:0] == []
+
+  # start=-5, end=Unbounded
+  list_[-5:] == [5, 6, 7, 8, 9]
+
+  # start=-5, end=0
+  list_[-5:0] == []
+
+
+  # start=0, end=5
+  list_[0:5] == [0, 1, 2, 3]
+
+  # start=Unbounded, end=5
+  list_[0:5] == [0, 1, 2, 3]
+
+  # start=0, end=-5
+  list_[0:-5] == [0, 1, 2, 3, 4]
+
+  # start=Unbounded, end=-5
+  list_[0:-5] == [0, 1, 2, 3, 4]
+
+
+  # start=2, end=6
+  list_[2:6] == [2, 3, 4, 5]
+
+  # start=2, end=-6
+  list_[2:-6] == [2, 3]
+
+  # start=-7, end=-2
+  list_[-7:-2] == [3, 4, 5, 6, 7]
+
+  // if using ltrim, this will remove all elements in the list
+  # start=-2, end=-6
+  list_[2:-6] == []
+
+  # empty list
+  list_[-1:1] == []
+ */
+message Unbounded { }
+
 message _ListFetchRequest {
   bytes list_name = 1;
   // Inclusive. 0 (start of list) by default.
   // A negative index counts from the end of the list
-  sint32 start_index = 2;
-  // If unset or 0, this effectively means list.length()
+  oneof start_index {
+    Unbounded unbounded_start = 2;
+    sint32 inclusive_start = 3;
+  }
+  // Exclusive if set.
+  // If unset, inclusive, this effectively means list.length()
   // If count is > the number of elements to return, return as much as you can
-  uint32 count = 3;
+  oneof end_index {
+    Unbounded unbounded_end = 4;
+    sint32 inclusive_end = 5;
+  }
+}
+
+message _ListRetainRequest {
+  bytes list_name = 1;
+  oneof start_index {
+    Unbounded unbounded_start = 2;
+    sint32 inclusive_start = 3;
+  }
```

Comment:
Yep, unfortunately, the thread was folded, but `start_index` will have a explicit presence so that we can support reverse order in the future

Reply:
So, we would be adding in the future, if asked, a numerical variable would be added, lets use `step` for now, this would control our traversal direction as well as controlling our stride, so not just reversing the list.

This would be a optional param and would not change any of the default behavior for existing workflows. But if `step` is included in the request and is negative and `start` is Unbounded, then `starts` becomes (list.len - 1)

But overall, I don't think making `start`  have a explicit presence hurt us, it might add to confusion as to why, but documentation here can cover that gap.

@malandis, anything to add? Also to double check what I got here is right 😄

Reply:
...i see.

Reply:
Yes that's right. In our zoom conversation we talked about the same thing but with start, count, and "direction of travel" (if I remember that right?). The `start_index` behavior we're talking about here will support that should we 
 implement it